### PR TITLE
[OpenMP][MLIR] Extend record member map support for omp dialect to LLVM-IR

### PR DESCRIPTION
**Pull request by agozillon**
This patch seeks to extend the current record type map support that was put in
place for Fortran's descriptor types to handle explicit member mapping for
record types.

This patch seeks to support this by extending the OpenMPToLLVMIRTranslation phase
to more generally support record types, building on the prior groundwork in the
Fortran allocatables/pointers patch. It now supports different kinds of record type
mapping, in this case full record type mapping and then explicit member mapping
in which there is a special case for certain types when mapped individually to not
require any parent map link in the kernel argument structure.

### CODE REVIEW

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2423]:
```suggestion
      cast<mlir::omp::MapInfoOp>(mapData.MapClause[mapDataIdx]);
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2461]:
```suggestion
  // If we have a partial map (no parent referenced in the map clauses of the
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2465]:
```suggestion
    auto memberClause = mlir::cast<mlir::omp::MapInfoOp>(
```

File: mlir/test/Target/LLVMIR/omptarget-fortran-allocatable-types-host.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/omptarget-fortran-allocatable-types-host.mlir:6]:
```suggestion
// alongside the omp.map.info, the test utilises mapping of array sections, 
```

File: mlir/test/Target/LLVMIR/omptarget-nested-record-type-mapping-host.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/omptarget-nested-record-type-mapping-host.mlir:70]: Nit: Trailing space in the last line.
```suggestion

```

File: mlir/test/Target/LLVMIR/omptarget-record-type-mapping-host.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/omptarget-record-type-mapping-host.mlir:63]: Nit: Trailing space in the last line.
```suggestion

```

File: mlir/test/Target/LLVMIR/omptarget-nested-record-type-mapping-host.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/omptarget-nested-record-type-mapping-host.mlir:5]:
```suggestion
// to LLVM-IR from MLIR when performing explicit member mapping of a record type
```

File: mlir/test/Target/LLVMIR/omptarget-record-type-mapping-host.mlir
Comment (skatrak) [inline mlir/test/Target/LLVMIR/omptarget-record-type-mapping-host.mlir:5]:
```suggestion
// to LLVM-IR from MLIR when performing explicit member mapping of a record type
```

**Review by skatrak — COMMENTED** [review]
Thank you Andrew, as usual I have a lot of nits but the overall approach seems reasonable to me.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (jsjodin) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2388]: Another dyn_cast here that could be a cast.

**Review by jsjodin — APPROVED** [review]
Other than fixing some nits I think this looks good.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [reply mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2103]: Disregard this suggestion, it's not the same. Sorry about that!

File: mlir/test/Target/LLVMIR/omptarget-record-type-mapping-host.mlir
Comment (skatrak) [reply mlir/test/Target/LLVMIR/omptarget-record-type-mapping-host.mlir:63]: Still there! 😉

**Review by skatrak — APPROVED** [review]
Thank you Andrew for working on my suggestions. LGTM, I just have some minimal nits, but there's no need for another review from me before merging this PR.

**Comment by agozillon** [comment]
I made a bit of a mistake not using SPR to land this PR series, so i've ended up having to directly commit the PR, the associated commit for this PR can be found here: https://github.com/llvm/llvm-project/commit/462435f089b7ca67ed1eed60f369e16b71e76b74

**Pull request merged**
